### ISSUE TITLE
[Plan] 面向 PCVR 排序的 unified token model with S-side prefix cache / cross-candidate KV cache

### ISSUE THREAD

**Issue by Puiching-Memory** [Puiching-Memory/TAAC_2026#88]
# [Epic] PCVR Unified Token Model with S-side Prefix Cache / Cross-candidate KV Cache

## 背景

TAAC 2026 主题强调“迈向统一序列建模与特征交互的大规模推荐系统”。当前 PCVR 模型大多按 row 逐样本打分，用户行为序列和候选 item 特征在每条样本中重复前向。若同一用户/请求下存在多个候选 item，序列侧重计算会被重复执行，推理延迟和吞吐存在优化空间。

本 issue 计划探索一个 PCVR 排序模型：训练时保持统一 token 化和统一 block，使用户行为序列、用户特征、item 特征在同一架构中深度交互；推理时将 item-independent 的 S-side 计算拆出并缓存，让不同候选 item 通过 cross-attention / KV cache 读取同一份用户侧中间态。

核心原则：这不是 late-fusion 的 `user_repr + item_mlp`，而是 unified token model 的 staged execution。

## 目标

- 构建一个新的 PCVR 实验包，例如 `experiments/prefix_cache_unified/`。
- 统一建模：
  - S tokens：多域用户行为序列。
  - NS tokens：user / item / context 非序列特征。
  - 统一 Transformer/HyFormer/OneTrans-style block 处理 S 与 NS token。
- 推理加速：
  - S-side prefix hidden states 或 per-layer K/V 可缓存。
  - 多个候选 item 复用同一用户侧 cache。
  - item tokens 仍通过 cross-attention 与用户序列 cache 做候选感知交互。
- 保持当前 PCVR runtime/checkpoint/bundle 契约兼容。
- 建立 AUC、吞吐、cache hit rate、延迟的消融记录。

## 非目标

- 不做 DFlash / LLM speculative decoding 依赖集成。
- 不引入 vLLM、SGLang、Transformers 大模型服务栈。
- 不把模型退化成简单双塔或固定 `user_repr` late fusion。
- 第一版不强求跨请求持久化 cache，先做 batch/request 内可验证复用。

### 1. 模型接口拆分

在实验模型内提供两段式接口：

```python
def encode_user(inputs: ModelInput) -> UserPrefixCache:
    ...

def score_item(inputs: ModelInput, user_cache: UserPrefixCache) -> torch.Tensor:
    ...

def forward(inputs: ModelInput) -> torch.Tensor:
    user_cache = self.encode_user(inputs)
    return self.score_item(inputs, user_cache)

def predict(inputs: ModelInput) -> tuple[torch.Tensor, torch.Tensor]:
    ...
```

`forward()` 仍保持普通端到端训练；`predict()` 后续可启用 cache-aware path。

### 2. V0：Hidden-state Cache

先缓存 S-side hidden states，而不是每层 K/V：

```text
S tokens -> S-side encoder -> seq_hidden_cache
item/user/context NS tokens -> item query tokens
item query tokens cross-attend seq_hidden_cache
classifier -> pCVR logit
```

优点：实现简单，便于验证 AUC 和吞吐收益。

### 3. V1：Per-layer KV Cache

在 V0 成立后升级：

```text
每层缓存 S-side K/V
item/NS tokens 在每层读取 cached K/V
保留统一 block 的深层交互语义
```

这是最贴近 cross-candidate KV cache 的版本。

### 4. Cache Key 策略

第一版优先支持 batch 内复用：

```text
cache_key = user_id + timestamp/request marker + seq_lens
```

若没有稳定 request id，则先统计 `user_id` 连续重复率和序列一致性，再决定是否加入 sequence hash。

### Phase 0: 数据与收益假设验证

- [ ] 增加/运行诊断：统计评估数据中同一 `user_id` 的连续重复次数、平均候选数、最大候选数。
- [ ] 统计相同 `user_id` 下序列字段是否完全一致或近似一致。
- [ ] 输出 cache hit potential 报告：`hit_rate`, `avg_candidates_per_user`, `p50/p90/p99 group size`。
- [ ] 如果数据基本随机散排，记录结论，优先保留模型结构实验，降低 cache 优先级。

### Phase 1: 新实验包 V0

- [ ] 新建 `experiments/prefix_cache_unified/__init__.py`。
- [ ] 新建 `experiments/prefix_cache_unified/model.py`。
- [ ] 复用现有 `ModelInput` 契约和 PCVR train/eval/infer runtime。
- [ ] 实现统一 tokenization：S tokens + user NS tokens + item NS tokens。
- [ ] 实现 `encode_user()` 输出 `UserPrefixCache`。
- [ ] 实现 `score_item()`，item tokens 对 cached S hidden states 做 cross-attention。
- [ ] `forward()` 保持无 cache 的训练路径，确保梯度正常。

### Phase 2: Cache-aware Predict Path

- [ ] 在 `predict()` 内实现 batch 内 user cache 复用。
- [ ] 增加配置开关：`prefix_cache_enabled`。
- [ ] 增加 cache 统计：命中数、miss 数、复用候选数、cache 构建耗时、score 耗时。
- [ ] 确保关闭 cache 时 logits 与普通路径一致或误差在可解释范围内。
- [ ] 避免在训练路径引入 Python dict 导致 compile/DDP 风险。

### Phase 3: Runtime / Benchmark

- [ ] 增加 focused benchmark：普通 row-wise predict vs cache-aware predict。
- [ ] 记录 batch size、AMP、device、num_workers、cache hit rate。
- [ ] 对比 `experiments/baseline`、`experiments/baseline_plus`、新实验 V0。
- [ ] 若 cache 只在特定排序下有效，补充数据排序策略或说明线上假设。

### Phase 4: V1 KV Cache

- [ ] 将 S-side hidden cache 扩展为 per-layer K/V cache。
- [ ] item/NS tokens 在每层执行 cross-attention against cached S K/V。
- [ ] 验证 V1 相比 V0 的 AUC/延迟收益。
- [ ] 评估 cache 内存占用，设置最大 cache entries 或 batch-only 限制。

## 验收标准

- [ ] 新实验包可被实验发现系统加载。
- [ ] 本地 smoke train/val/infer 跑通。
- [ ] checkpoint sidecar 可用于评估和推理重建模型。
- [ ] `prefix_cache_enabled=false` 时行为稳定，可作为模型结构 baseline。
- [ ] `prefix_cache_enabled=true` 时输出预测文件格式不变。
- [ ] 给出至少一份 benchmark 报告：
  - AUC 或本地验证指标。
  - rows/s 或 samples/s。
  - p50/p90 batch latency。
  - cache hit rate。
  - 显存占用。
- [ ] 技术说明中明确：cache 是 unified model 的 staged execution，不是 late fusion。

## 需要关注的代码区域

- `experiments/<new_experiment>/__init__.py`
- `experiments/<new_experiment>/model.py`
- `src/taac2026/application/evaluation/workflow.py`
- `src/taac2026/domain/model_contract.py`
- `tests/unit/experiments/test_packages.py`
- `tests/unit/experiments/test_runtime_contract_matrix.py`

## 测试计划

```bash
uv run pytest tests/unit/experiments/test_packages.py -q
uv run pytest tests/unit/experiments/test_runtime_contract_matrix.py -q
uv run pytest tests/unit/domain/test_model_contract.py -q
uv run pytest tests/unit/application/experiments -q
```

若修改共享预测 runtime，再补充对应 application/evaluation focused tests。

## 风险

- 官方评估数据若不按用户/request 聚集，cache hit rate 可能不足。
- 过早把用户序列压成固定向量会损害 candidate-aware 交互，偏离主题。
- Per-layer KV cache 可能增加显存占用，吞吐收益需要实测。
- Python 侧 cache 管理可能影响 `torch.compile` 或线上稳定性。
- 训练/推理路径分叉后，必须防止 checkpoint 和 sidecar 契约漂移。

## 开放问题

- 评估 parquet 是否存在 request id 或天然按 `user_id` 聚集？
- 同一用户多候选时，序列字段是否完全相同？
- cache key 应使用 `user_id + timestamp`，还是需要 sequence fingerprint？
- 第一版目标更偏 AUC，还是更偏延迟合规和吞吐？
- 新实验应基于 Baseline HyFormer、OneTrans，还是 Symbiosis 的部分组件裁剪？
